pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -177,3 +158,304 @@ func wrapServerToClientError(err error) error {
 	}
 	return newErrorf(codeBackendDisconnected, "copying from target server to client: %s", err)
 }
+
+// tryReportError tries to send err to errCh, and closes the forwarder if
+// it succeeds. If an error has already been reported, err will be dropped.
+func (f *forwarder) tryReportError(err error) {
+	select {
+	case f.errCh <- err: /* error reported */
+		f.Close()
+	default: /* the channel already contains an error */
+	}
+}
+
+// aLongTimeAgo is a non-zero time, far in the past, used for immediate
+// cancellation of dials.
+var aLongTimeAgo = timeutil.Unix(1, 0)
+
+var errProcessorResumed = errors.New("processor has already been resumed")
+
+type processor interface {
+	// resume starts the processor and blocks if it hasn't been started yet.
+	// This returns nil if the processor can be resumed again in the future.
+	// If an error (with the exception of errProcessorResumed) was returned,
+	// the processor should not be resumed again, and the forwarder should be
+	// closed.
+	resume() error
+	// suspend requests for the processor to be suspended if it is in a safe
+	// state. Callers must call waitUntilSuspended to ensure that the processor
+	// has been suspended. This is because there's a chance where the processor
+	// is not in a state to be suspended (e.g. blocked on forwarding a large
+	// message). If the suspend request failed, an error will be returned. The
+	// caller is safe to retry again.
+	suspend() error
+	// waitUntilSuspended blocks until the processor has been suspended. If
+	// suspend was not called before, this may block forever.
+	waitUntilSuspended()
+}
+
+// requestProcessor handles the communication from the client to the server.
+type requestProcessor struct {
+	f  *forwarder
+	mu struct {
+		syncutil.Mutex
+		cond       *sync.Cond
+		resumed    bool
+		inPeek     bool
+		suspendReq bool // Indicates that a suspend has been requested.
+	}
+	wg sync.WaitGroup
+}
+
+var _ processor = &requestProcessor{}
+
+func (p *requestProcessor) lock() {
+	p.mu.Lock()
+	if p.mu.cond == nil {
+		p.mu.cond = sync.NewCond(&p.mu)
+	}
+}
+
+func (p *requestProcessor) unlock() { p.mu.Unlock() }
```

Comment:
The fact there are two different requestProcessor implementations looks like an artifact of the fact BackendConn and FrontendConn are different types. Can we merge the two requestProcessor implementations if we pass them a pgInterceptor instead of the BackendConn and FrontendConn?

Reply:
Possibly, but I definitely do not want to do that in this PR. That change is not trivial, and requires more design thought. Also note that this structure looks similar for now, but with the connection migration work in, it will be different because the processors will need to do different operations (e.g. setting various variables), though you can argue that some kind of callbacks may work here.